resources/aws_ec2.go

```diff
@@ -912,143 +822,149 @@ func (obj *AwsEc2Res) prependName() string {
-// or the other.
-func waitUntilInstanceStoppedOrTerminatedWithContext(ctx context.Context, waitInput *ec2.DescribeInstancesInput, c *ec2.EC2) error {
-	errChan := make(chan error)
-	defer close(errChan) // unnecessary, but nice to have
-	wg := sync.WaitGroup{}
-	defer wg.Wait()
+	// if we got one, return it.
+	return diOutput.Reservations[0].Instances[0], nil
+}
 
-	closeChan := make(chan struct{})
-	innerCtx, cancel := context.WithCancel(context.TODO()) // uncoupled!!
+// longpollWaiter waits for an instance to change state, and returns the state.
+func longpollWaiter(ctx context.Context, instance *ec2.Instance, c *ec2.EC2) (string, error) {
+	var err error
+	var name string
 
-	once := &sync.Once{}
-	closer := func() {
-		cancel()
-		close(closeChan)
+	// these cases are not permitted.
+	if instance == nil {
+		return "", fmt.Errorf("nil instance")
+	}
+	if aws.StringValue(instance.State.Name) == "" {
+		return "", fmt.Errorf("nil or empty state")
 	}
-	defer once.Do(closer) // needed if we exit below due to ctx being cancelled
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		defer once.Do(closer)
 
-		err := c.WaitUntilInstanceStoppedWithContext(innerCtx, waitInput)
-		if err != nil {
-			if aerr, ok := err.(awserr.Error); ok {
-				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
-					err = nil // we want to ignore these kinds of errors
-				}
-			}
+	// get the instance name.
+	for _, tag := range instance.Tags {
+		if aws.StringValue(tag.Key) == "Name" {
+			name = aws.StringValue(tag.Value)
 		}
+	}
+	// error if we didn't find one.
+	if name == "" {
+		return "", fmt.Errorf("name not found")
+	}
 
-		select {
-		case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to stop"):
-		case <-closeChan:
-		}
-	}()
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		defer once.Do(closer)
-
-		err := c.WaitUntilInstanceTerminatedWithContext(innerCtx, waitInput)
-		if err != nil {
-			if aerr, ok := err.(awserr.Error); ok {
-				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
-					err = nil
-				}
+	// instance.InstanceId might be nil in case of terminated instances, which
+	// the api will treat as if it is't included.
+	waitInput := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{instance.InstanceId},
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(name)},
+			},
+		},
+	}
+	// call on the appropriate waiter based on current state.
+	switch aws.StringValue(instance.State.Name) {
+	case ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping:
+		err = c.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameStopped, ec2.InstanceStateNamePending:
+		err = c.WaitUntilInstanceRunningWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameTerminated:
+		err = c.WaitUntilInstanceExistsWithContext(ctx, waitInput)
```

Comment:
fwiw, there's a race where this returns some milliseconds before the instance is available, so we get 4 or 5 extra "terminated" events. I know sleeping for a second isn't a great fix, but it would work in this case.